nhs/data/allocation.py

```diff
@@ -0,0 +1,286 @@
+    respectively will be included in the output.
+
+    **Note**: If `left_code_col` and `right_code_col` have different data types, then `left_code_col`
+    in `census` will be type-casted to the data type of `right_code_col` in `coords`. Output will
+    also use `left_code_col` if `left_code_col` and `right_code_col` have different names.
+
+    Parameters
+    ----------
+    census : pl.LazyFrame
+        The census dataframe to join.
+    coords : pl.LazyFrame
+        The gnaf coordinate dataframe to join.
+    left_code_col : str
+        The column in the census dataframe to join on.
+    right_code_col : str
+        The column in the gnaf coordinate dataframe to join on.
+
+    Returns
+    -------
+    pl.LazyFrame
+        A LazyFrame containing the column `right_code_col` and the rest of the columns
+        from `census` and `coords`.
+
+    Examples
+    --------
+    >>> census = pl.LazyFrame({
+    ...     "CODE_2021": ["123", "45", "3"],
+    ...     "feature1": [24, 65, 234]
+    ... })
+    >>> coords = pl.LazyFrame({
+    ...     "CODE21": [123, 45, 3],
+    ...     "longitude": [134.56, 456.12, 21.124],
+    ...     "latitude": [-34.56, -23.12, 12.124],
+    ... })
+    >>> join_census_with_coords(
+    ...     census, coords, left_code_col="CODE_2021", right_code_col="CODE21"
+    ... ).collect()
+        shape: (3, 4)
+        ┌───────────┬──────────┬───────────┬──────────┐
+        │ CODE_2021 ┆ feature1 ┆ longitude ┆ latitude │
+        │ ---       ┆ ---      ┆ ---       ┆ ---      │
+        │ i64       ┆ i64      ┆ f64       ┆ f64      │
+        ╞═══════════╪══════════╪═══════════╪══════════╡
+        │ 123       ┆ 24       ┆ 134.56    ┆ -34.56   │
+        │ 45        ┆ 65       ┆ 456.12    ┆ -23.12   │
+        │ 3         ┆ 234      ┆ 21.124    ┆ 12.124   │
+        └───────────┴──────────┴───────────┴──────────┘
+    """
+    census_type_casted = census.with_columns(
+        pl.col(left_code_col).cast(coords.collect_schema()[right_code_col])
+    )
+
+    return census_type_casted.join(
+        coords, left_on=left_code_col, right_on=right_code_col, how="inner"
+    )
+
+
+@log_entry_exit()
+def sample_census_feature(
+    census: pl.LazyFrame, code_col: str, long_col: str, lat_col: str, feature_col: str
+):
+    """
+    Randomly sample rows from each group in the census LazyFrame using value of `feature_col` as sample size.
+
+    Parameters
+    ----------
+    census: pl.LazyFrame
+        A `LazyFrame` to sample from, with columns `code_col`, `long_col`, `lat_col`,
+        and `feature_col`.
+    code_col : str
+        Column in `census` that contains some identifier/code.
+    long_col : str
+        Column in `census` that contains the longitude.
+    lat_col : str
+        Column in `census` that contains the latitude.
+    feature_col : str
+        Column in `census` for a feature where each row is the
+        sample size for each group.
```

Comment:
I agree with these inputs for now - the proposed changes will affect this, chaning code_col to the 'category' name which we will have a dicitionary mapping for each that will return ALL associated column features. 

And is it the correct interpretation that feature_col belongs to the second variables and is where we get the totals / proportions etc... that will also be changed.

Reply:
Not sure what you mean by "second variables", `feature_col` is a column in the `census` lazy frame, that column will have the number of people with that feature per SA1 region